src/data/pfunctor/multivariate/W.lean

```diff
@@ -0,0 +1,253 @@
+  P.W α → C
+| ⟨a, f'⟩ :=
+  let g' (a : P.A) (f : P.last.B a → P.last.W) (h : P.W_path ⟨a, f⟩ ⟹ α)
+        (h' : P.last.B a → C) : C :=
+      g a (P.W_path_dest_left h) (λ i, ⟨f i, P.W_path_dest_right h i⟩) h' in
+  P.Wp_rec g' a f'
+
+theorem W_rec_eq {α : typevec n} {C : Type*}
+    (g : Π a : P.A, ((P.drop).B a ⟹ α) → ((P.last).B a → P.W α) → ((P.last).B a → C) → C)
+    (a : P.A) (f' : P.drop.B a ⟹ α) (f : P.last.B a → P.W α) :
+  P.W_rec g (P.W_mk a f' f) = g a f' f (λ i, P.W_rec g (f i)) :=
+begin
+  rw [W_mk, W_rec], dsimp, rw [Wp_rec_eq],
+  dsimp only [W_path_dest_left_W_path_cases_on, W_path_dest_right_W_path_cases_on],
+  congr; ext1 i; cases (f i); refl
+end
+
+theorem W_ind {α : typevec n} {C : P.W α → Prop}
+    (ih : ∀ (a : P.A) (f' : P.drop.B a ⟹ α) (f : P.last.B a → P.W α),
+      (∀ i, C (f i)) → C (P.W_mk a f' f)) :
+  ∀ x, C x :=
+begin
+  intro x, cases x with a f,
+  apply @Wp_ind n P α (λ a f, C ⟨a, f⟩), dsimp,
+  intros a f f' ih',
+  dsimp [W_mk] at ih,
+  let ih'' := ih a (P.W_path_dest_left f') (λ i, ⟨f i, P.W_path_dest_right f' i⟩),
+  dsimp at ih'', rw W_path_cases_on_eta at ih'',
+  apply ih'',
+  apply ih'
+end
+
+theorem W_cases {α : typevec n} {C : P.W α → Prop}
+    (ih : ∀ (a : P.A) (f' : P.drop.B a ⟹ α) (f : P.last.B a → P.W α), C (P.W_mk a f' f)) :
+  ∀ x, C x :=
+P.W_ind (λ a f' f ih', ih a f' f)
+
+/-- W-types are functorial -/
+def W_map {α β : typevec n} (g : α ⟹ β) : P.W α → P.W β :=
+λ x, g <$$> x
+
+theorem W_mk_eq {α : typevec n} (a : P.A) (f : P.last.B a → P.last.W)
+    (g' : P.drop.B a ⟹ α) (g : Π j : P.last.B a, P.W_path (f j) ⟹ α) :
+  P.W_mk a g' (λ i, ⟨f i, g i⟩) =
+    ⟨⟨a, f⟩, P.W_path_cases_on g' g⟩ := rfl
+
+theorem W_map_W_mk {α β : typevec n} (g : α ⟹ β)
+    (a : P.A) (f' : P.drop.B a ⟹ α) (f : P.last.B a → P.W α) :
+  g <$$> P.W_mk a f' f = P.W_mk a (g ⊚ f') (λ i, g <$$> f i) :=
+begin
+  show _ = P.W_mk a (g ⊚ f') (mvfunctor.map g ∘ f),
+  have : mvfunctor.map g ∘ f = λ i, ⟨(f i).fst, g ⊚ ((f i).snd)⟩,
+  { ext i : 1, dsimp [function.comp], cases (f i), refl },
+  rw this,
+  have : f = λ i, ⟨(f i).fst, (f i).snd⟩,
+  { ext1, cases (f x), refl },
+  rw this, dsimp,
+  rw [W_mk_eq, W_mk_eq],
+  have h := mvpfunctor.map_eq P.Wp g,
+  rw [h, comp_W_path_cases_on]
+end
+
+-- TODO: this technical theorem is used in one place in constructing the initial algebra.
+-- Can it be avoided?
+
+/-- Constructor of a value of `P.obj (α ::: β)` from components.
+Useful to avoid complicated type annotation -/
+@[reducible] def obj_append1 {α : typevec n} {β : Type*}
+    (a : P.A) (f' : P.drop.B a ⟹ α) (f : P.last.B a → β) :
+  P.obj (α ::: β) :=
+⟨a, split_fun f' f⟩
+
+theorem map_obj_append1 {α γ : typevec n} (g : α ⟹ γ)
+  (a : P.A) (f' : P.drop.B a ⟹ α) (f : P.last.B a → P.W α) :
+append_fun g (P.W_map g) <$$> P.obj_append1 a f' f =
+  P.obj_append1 a (g ⊚ f') (λ x, P.W_map g (f x)) :=
+by rw [obj_append1, obj_append1, map_eq, append_fun, ← split_fun_comp]; refl
+
+/-
```

Comment:
Maybe some of these comments should be module docs `/-! ... -/` so they appear in the docs?